### PR TITLE
Issue 27 what spi

### PR DESCRIPTION
**Pull request by lawik**
Not sure if we have a good way of testing the chunking stuff since that's deep in the IO layer. Any thoughts?

Aside from that, would be good to have your eye on this before pushing it through. Probably makes sense to make 1.0.1 for this right?

### CODE REVIEW

**Review by BrynjarM — APPROVED** [review]
Tested and verified (ee68044fd3424cbd836034b8841229804e152f27) that it works  on my wHAT

File: lib/hal/rpiio.ex [hunk 1]
```diff
@@ -34,6 +34,7 @@ defmodule Inky.RpiIO do
   @spi_speed_hz 488_000
   @spi_command 0
   @spi_data 1
+  @spi_chunk_size 4096
```
Comment (nyaray) [inline lib/hal/rpiio.ex]: Should be suffixed with `bits` or `bytes` for clarity of intent.

File: lib/hal/rpiio.ex [hunk 2]
```diff
@@ -93,12 +94,28 @@ defmodule Inky.RpiIO do
 
   defp spi_write(pins, data_or_command, values) when is_list(values) do
     :ok = GPIO.write(pins.dc_pid, data_or_command)
-    {:ok, <<_::binary>>} = SPI.transfer(pins.spi_pid, :erlang.list_to_binary(values))
+    spi_write(pins, data_or_command, :erlang.list_to_binary(values))
```
Comment (nyaray) [inline lib/hal/rpiio.ex]: Does it not break tests that you added an unnecessary duplicated write to `dc`? Remove the GPIO write if this is going to rely on the later clause via the recursive call.

File: lib/hal/rpiio.ex [hunk 3]
```diff
@@ -93,12 +94,28 @@ defmodule Inky.RpiIO do
...
+        size = byte_size(value)
+        parts = div(size - 1, @spi_chunk_size)
+
+        for x <- 0..parts do
+          offset = x * @spi_chunk_size
+          left = size - offset
+          length = if left > @spi_chunk_size, do: @spi_chunk_size, else: left
+
+          {:ok, <<_::binary>>} = SPI.transfer(pins.spi_pid, :binary.part(value, offset, length))
+        end
```
Comment (nyaray) [inline lib/hal/rpiio.ex]: * lines 108-117 should probably be in a separate function, named something like `spi_write_chunked`
* `x` should be named `chunk_index` and `parts`, `chunk_count`

File: lib/hal/rpihal.ex [hunk 4]
```diff
@@ -158,14 +159,15 @@ defmodule Inky.RpiHAL do
...
 
-  defp configure_if_yellow(state, accent) do
-    # Set voltage of VSH and VSL on Yellow device
-    if accent == :yellow do
-      write_command(state, 0x04, 0x07)
-    else
-      state
-    end
-  end
+  # Set voltage of VSH and VSL on Yellow device
```
Comment (nyaray) [inline lib/hal/rpihal.ex:165]: The comment should probably be copied to the red what function.

**Review by nyaray — CHANGES_REQUESTED** [review]

File: lib/hal/rpiio.ex (hunk 2 above)
Comment (nyaray) [reply lib/hal/rpiio.ex]: My bad, upon closer inspection, we didn't have a test for that.

File: lib/hal/rpihal.ex (hunk 4 above)
Comment (nyaray) [reply lib/hal/rpihal.ex:165]: ✓

**Review by nyaray — APPROVED** [review]

**Comment by nyaray** [comment]
@BrynjarM would you mind verifying the latest commit on this branch?

Since you've got a wHAT, I'll also be asking you to check out a performance improvement I've been working at some point in the not too distant future, if that's ok!

**Comment by BrynjarM** [comment]
@nyaray sorry for the delay, tested fe9bab3e268605ee39c54e097b31576915f94765 and it works on my wHAT

Sure, just ping me when the performance improvements are ready

**Comment by JonRowe** [comment]
Any chance of releasing this 😬 (sorry to ask but I ran into the what bug too and then noticed it was fixed here 😂 )

**Comment by lawik** [comment]
Ah, I'll take a look :)

**Pull request merged**